src/Xrefcheck/Scan.hs

```diff
@@ -67,35 +66,58 @@ data ExclusionConfig' f = ExclusionConfig
 
 makeLensesWith postfixFields ''ExclusionConfig'
 
-normaliseExclusionConfigFilePaths :: ExclusionConfig -> ExclusionConfig
-normaliseExclusionConfigFilePaths ec@ExclusionConfig{..}
-  = ec
-    { ecIgnore            = map normaliseGlobPattern ecIgnore
-    , ecIgnoreLocalRefsTo = map normaliseGlobPattern ecIgnoreLocalRefsTo
-    , ecIgnoreRefsFrom    = map normaliseGlobPattern ecIgnoreRefsFrom
-    }
-
 -- | File extension, dot included.
 type Extension = String
 
 -- | Way to parse a file.
-type ScanAction = FilePath -> IO (FileInfo, [ScanError])
+type ScanAction = CanonicalPath -> IO (FileInfo, [ScanError 'Parse])
 
 -- | All supported ways to parse a file.
 type FormatsSupport = Extension -> Maybe ScanAction
 
 data ScanResult = ScanResult
-  { srScanErrors :: [ScanError]
+  { srScanErrors :: [ScanError 'Gather]
   , srRepoInfo   :: RepoInfo
-  } deriving stock (Show)
+  }
 
-data ScanError = ScanError
-  { sePosition    :: Position
-  , seFile        :: FilePath
+-- | A scan error indexed by different process stages.
+--
+-- Within 'Parse', 'seFile' has no information because the same
+-- file is being parsed.
+--
+-- Within 'Gather', 'seFile' stores the 'FilePath' corresponding
+-- to the file in where the error was found.
+data ScanError (a :: ScanStage) = ScanError
```

Comment:
Do you think that this 'Trees that grow' style fits in here?

Reply:
Oh, I like your type families solution here, it makes particular sense. At your place, I would probably write it in the same way :)

However I think that two separate datatypes would work better here, and here's why: our core part serves as sort of API for potential scanners developers, and later there will be a lot of code that works with `ScanError 'Parse`.

And with the current approach, we win at the core side (fewer number of datatypes) and lose at the parsers' side (one has to always initialize that `seFile = ()`), and the latter is more significant for the code amount and API-wise.

Reply:
Would it be an option to use `mkParseScanError ` instead of initialising `seFile` to `()`? I can also give type aliases to `ScanError 'Parse` and `ScanError 'Gather` also.

Reply:
Oh right, we already use a smart constructor there, so avoiding `()` each time won't be a problem.

Okay, let's go with it.